src/utilization.rs

```diff
@@ -96,25 +96,25 @@ impl Timer {
     }
 
     /// Begin a new span representing time spent waiting
-    fn start_wait(&mut self) {
+    pub fn start_wait(&mut self) {
         if !self.waiting {
             self.end_span();
             self.waiting = true;
         }
     }
 
     /// Complete the current waiting span and begin a non-waiting span
-    fn stop_wait(&mut self) {
-        assert!(self.waiting);
-
-        self.end_span();
-        self.waiting = false;
+    pub fn stop_wait(&mut self) {
+        if self.waiting {
+            self.end_span();
+            self.waiting = false;
+        }
```

Comment:
Could this model be changed to eliminate the need for the `waiting` flag and mutability in `start_wait`? What I'm thinking is to have `start_wait` return a newtype wrapping `Instant` that is consumed by `stop_wait`, effectively encoding the `waiting` flag in the type/ownership system. This will require reordering the send/receive in the loops in topology, but I think it should be doable if the output buffer is moved out of the loop.

Reply:
Yeah this is a good idea. I'll plan to move in that direction in the next PR, since it meshes with some of the reporting pattern changes. For now I'll likely just leave it in favor of getting this step merged.